### PR TITLE
ruby class layout style

### PR DESCRIPTION
**Pull request by soma**
See discussion here:

https://github.com/barsoom/auctionet/commit/a5cfbe6e028f6b4252cd810cab6893ea5517ea0d#commitcomment-9044301

This adds a rule that make SENSE.

- [x] AR
- [x] TS
- [x] HN
- [x] JK

(Not Kim since she won't be back within 2 weeks.)

### CODE REVIEW

File: styleguide/README.md [hunk 1]
```diff
@@ -157,6 +157,49 @@ Use your own judgment to choose between the options.
 ---
 ### Ruby
 
+### Use a consistent class layout
```
Comment (henrik) [inline styleguide/README.md]: Four hashes: `####`

File: styleguide/README.md [hunk 2]
```diff
@@ -157,6 +157,49 @@ Use your own judgment to choose between the options.
...
+
+See [Ruby Style Guide on class layout](https://github.com/bbatsov/ruby-style-guide#consistent-classes)
```
Comment (henrik) [inline styleguide/README.md]: Do we need to reference this? I'm not a fan of bbatsov's styleguide.

Comment (soma) [inline styleguide/README.md]: We do if we copy paste stuff like I just did!

Comment (henrik) [inline styleguide/README.md]: Ah, maybe say "Based on …" then :) I don't want us to reference that styleguide as an authority.

**Comment by henrik** [comment]
This order matches what I like. It's a :thumbsup: from me, though I have some non-blocking concerns:

* I'd like us to clarify that we only reference bbatsov as a source, not an authority.
* The comment style (all lowercase, no punctuation) look sloppy to me, but we don't regulate that.

Maybe we should be explicit that "attribute macros" include `pattr_initialize` and friends, since that was part of the debate?

**Comment by soma** [comment]
Feel free to fix the comments if you want, its just a copy paste from the eminent bbatsov.

**Comment by henrik** [comment]
I can't change them in `soma/master`.

**Comment by dentarg** [comment]
Do you have any preference around ordering inside the different groups (a group being e.g. public instance methods)? Alphabetically perhaps? Or is it arbitrary?

**Comment by soma** [comment]
I would prefer not having a rule for that, mostly because the order could be context bound, like grouping them after how they are used in the class. Classes should be short enough for this to not matter much :)

**Comment by henrik** [comment]
One group I like to order: Active Record callbacks (and similar things) – I like to order them in lifecycle order, e.g. have `before_create`, `after_create` in that order.

**Comment by joakimk** [comment]
This seems to be mostly about how a AR model is done. Some special rules may apply there because of the excessive use of macros.

In smaller classes I like to see an order that goes from highest level of abstraction to lowest level.

```ruby
class Car
  def self.drive
    engine = Engine.new
    new(engine).drive
  end

  pattr_initialize :engine

  def drive
    return if moving?
    turn_key
  end

  def moving?
  end

  private

  def turn_key
  end
end
```

The intended use for this class is `Car.drive`, but in unit tests you might want to do `.new(fake_engine)` and that's fine.

**Comment by henrik** [comment]
I think we're getting bogged down in fixing Everything in this single PR :) @tskogberg, @joakimk are you fine with the PR as it is written currently? And you are then free to make further PRs to add even more rules?

As I said, it's a :thumbsup: from me though I have those small concerns that I'd like to be fixed (I can't edit the PR code).

**Pull request merged**